ddl/ddl_api.go

```diff
@@ -58,39 +58,54 @@ import (
 
 const expressionIndexPrefix = "_V$"
 
-func (d *ddl) CreateSchema(ctx sessionctx.Context, schema model.CIStr, charsetInfo *ast.CharsetOpt) (err error) {
+func (d *ddl) CreateSchema(ctx sessionctx.Context, schema model.CIStr, charsetInfo *ast.CharsetOpt) error {
+	dbInfo := &model.DBInfo{Name: schema}
+	if charsetInfo != nil {
+		dbInfo.Charset = charsetInfo.Chs
+		dbInfo.Collate = charsetInfo.Col
+	} else {
+		dbInfo.Charset, dbInfo.Collate = charset.GetDefaultCharsetAndCollate()
+	}
+	return d.CreateSchemaWithInfo(ctx, dbInfo, OnExistError, false /*tryRetainID*/)
+}
+
+func (d *ddl) CreateSchemaWithInfo(
+	ctx sessionctx.Context,
+	dbInfo *model.DBInfo,
+	onExist OnExist,
+	tryRetainID bool,
+) error {
 	is := d.GetInfoSchemaWithInterceptor(ctx)
-	_, ok := is.SchemaByName(schema)
+	_, ok := is.SchemaByName(dbInfo.Name)
 	if ok {
-		return infoschema.ErrDatabaseExists.GenWithStackByArgs(schema)
+		err := infoschema.ErrDatabaseExists.GenWithStackByArgs(dbInfo.Name)
+		switch onExist {
+		case OnExistIgnore:
+			ctx.GetSessionVars().StmtCtx.AppendNote(err)
+			return nil
+		case OnExistError, OnExistReplace:
+			// FIXME: can we implement MariaDB's CREATE OR REPLACE SCHEMA?
+			return err
+		}
 	}
 
-	if err = checkTooLongSchema(schema); err != nil {
+	if err := checkTooLongSchema(dbInfo.Name); err != nil {
 		return errors.Trace(err)
 	}
 
-	genIDs, err := d.genGlobalIDs(1)
-	if err != nil {
+	if err := checkCharsetAndCollation(dbInfo.Charset, dbInfo.Collate); err != nil {
 		return errors.Trace(err)
 	}
-	schemaID := genIDs[0]
-	dbInfo := &model.DBInfo{
-		Name: schema,
-	}
 
-	if charsetInfo != nil {
-		err = checkCharsetAndCollation(charsetInfo.Chs, charsetInfo.Col)
-		if err != nil {
-			return errors.Trace(err)
-		}
-		dbInfo.Charset = charsetInfo.Chs
-		dbInfo.Collate = charsetInfo.Col
-	} else {
-		dbInfo.Charset, dbInfo.Collate = charset.GetDefaultCharsetAndCollate()
+	// FIXME: support `tryRetainID`.
+	genIDs, err := d.genGlobalIDs(1)
```

Comment:
Why not support it now? it is not complicated

Reply:
it may be easy but i don't wanna touch code beyond `sendDDLJob` here, since i'm not too familiar with that 🙃